ray_tile_and_stitch_util.py

```diff
@@ -0,0 +1,341 @@
+            ):
+                continue
+
+            # stack three bands into one array
+            img = np.stack((band_1_array, band_2_array, band_3_array), axis=2)
+            cv2.normalize(img, img, 0, 255, cv2.NORM_MINMAX)
+            img = img.astype(np.uint8)
+            B, G, R = cv2.split(img)
+            out_B = cv2.equalizeHist(B)
+            out_R = cv2.equalizeHist(R)
+            out_G = cv2.equalizeHist(G)
+            final_image = np.array(cv2.merge((out_B, out_G, out_R)))
+
+            # Upper left (x,y) for each images
+            ul_row_divided_img = uly + i * y_resolution
+            ul_col_divided_img = ulx + j * x_resolution
+
+            tile_metadata = ImageTileMetadata(
+                upper_left_row=ul_row_divided_img, upper_left_col=ul_col_divided_img, tile_num=tile_count, id_i=id_i, id_j=id_j)
+            image_tile = ImageTile(
+                tile_values=final_image, tile_metadata=tile_metadata)
+            tiles.append(image_tile)
+            tile_count += 1
+
+    # --------------- Store all the title as an object file
+    image_metadata = ImageMetadata(
+        len_x_list=len(x_list), len_y_list=len(y_list), x_resolution=x_resolution, y_resolution=y_resolution)
+    row["image_metadata"] = image_metadata
+    new_rows = []
+    tile_count = 0
+    for image_tile in tiles:
+        new_row = copy.deepcopy(row)
+        new_row["image_tile"] = image_tile
+        new_row["tile_num"] = tile_count
+        new_rows.append(new_row)
+        tile_count += 1
+    return new_rows
+
+
+def stitch_shapefile(group: pd.DataFrame):
+    """
+    Create a shapefile for each image.
+    Note that normally ray rows are dictionaries but this is a group because we've called
+    grouped by. When ray does a groupby and the rows can't be represented in numpy arrays
+    it uses pandas dataframe.
+
+    Parameters
+    ----------
+    group: a pandas dataframe that has all of the shapefile information for each tile in the
+            image.
+    """
+    image_shapefile_results = []
+    temp_polygon_dict = defaultdict(dict)
+    dict_ij = defaultdict(dict)
+    sorted_group = group.sort_values(by="tile_num")
+    for index, row in sorted_group.iterrows():
+        image_shapefile_results.extend(
+            row["tile_shapefile_results"].shapefile_results)
+        image_tile = row["image_tile"]
+        tile_num = row["tile_num"]
+        temp_polygon_dict[tile_num] = row["num_polygons_in_tile"]
+        id_i = image_tile.tile_metadata.id_i
+        id_j = image_tile.tile_metadata.id_j
+        dict_ij[id_i][id_j] = tile_num
+
+    polygon_dict = defaultdict(dict)
+    poly_count = 0
+    for k, v in temp_polygon_dict.items():
+        polygon_dict[k] = [poly_count, poly_count + v]
+        poly_count += v
+
+    first_row = group.head(1)
+    image_data_from_arbitrary_row = first_row["image_metadata"][0]
+    size_i, size_j = image_data_from_arbitrary_row.len_y_list, image_data_from_arbitrary_row.len_x_list
```

Comment:
Do we choose an arbitrary row because they all have the same value and it doesn't matter where we read it from?

Reply:
yes because we grouped by image, all the rows (each corresponding to a specific tile) should have the same image metadata